phylanx/ir/ranges.hpp

```diff
@@ -0,0 +1,155 @@
+            std::input_iterator_tag,
+            execution_tree::primitive_argument_type>
+    {
+    private:
+        using int_range_type = std::pair<std::int64_t, std::int64_t>;
+        using arg_range_type =
+            std::vector<execution_tree::primitive_argument_type>::iterator;
+        using iterator_type = util::variant<int_range_type, arg_range_type>;
+
+    public:
+        range_iterator(std::int64_t start, std::int64_t step)
+          : it_(std::make_pair(start, step))
+        {
+        }
+
+        range_iterator(arg_range_type it)
+          : it_(it)
+        {
+        }
+
+    private:
+        friend class hpx::util::iterator_core_access;
+
+        execution_tree::primitive_argument_type dereference() const;
+        bool equal(range_iterator const& other) const;
+        void increment();
+
+    private:
+        iterator_type it_;
+    };
+
+    //////////////////////////////////////////////////////////////////////////
+    class PHYLANX_EXPORT reverse_range_iterator
+        : public hpx::util::iterator_facade<reverse_range_iterator,
+        execution_tree::primitive_argument_type,
+        std::input_iterator_tag,
+        execution_tree::primitive_argument_type>
+    {
+    private:
+        using int_range_type = std::pair<std::int64_t, std::int64_t>;
+        using arg_range_type = std::vector<
+            execution_tree::primitive_argument_type>::reverse_iterator;
+        using iterator_type = util::variant<int_range_type, arg_range_type>;
+
+    public:
+        reverse_range_iterator(std::int64_t reverse_start, std::int64_t step)
+            : it_(std::make_pair(reverse_start, step))
+        {
+        }
+
+        reverse_range_iterator(arg_range_type it)
+            : it_(it)
+        {
+        }
+
+    private:
+        friend class hpx::util::iterator_core_access;
+
+        execution_tree::primitive_argument_type dereference() const;
+        bool equal(reverse_range_iterator const& other) const;
+        void increment();
+
+    private:
+        iterator_type it_;
+    };
+
+    //////////////////////////////////////////////////////////////////////////
+    class PHYLANX_EXPORT range
+    {
+    public:
+        using int_range_type =
+            hpx::util::tuple<std::int64_t, std::int64_t, std::int64_t>;
+        using args_type = std::vector<execution_tree::primitive_argument_type>;
+        using arg_pair_type = std::pair<range_iterator, range_iterator>;
+        using reverse_arg_pair_type =
+            std::pair<reverse_range_iterator, reverse_range_iterator>;
+        using range_type = util::variant<
+            int_range_type,
+            args_type,
+            arg_pair_type,
+            reverse_arg_pair_type>;
+
+        range_iterator begin();
+        range_iterator end();
+
+        reverse_range_iterator rbegin();
+        reverse_range_iterator rend();
+
+        explicit range(
+            std::vector<execution_tree::primitive_argument_type> data)
+          : data_(data)
```

Comment:
This should be `std::move(data)`. Alternatively you might want to provide two constructors, one for `std::vector<execution_tree::primitive_argument_type> const& data` and one for `std::vector<execution_tree::primitive_argument_type> && data`.

Reply:
Done.